File: spa-next/src/lib/featureFlags.js
// wyrd-0gou: SPA-side feature-flag mapping.
//
// The server ships the resolved config on /api/manifest:
//   config = { all: bool, flags: { <ENVIFIED_NAME>: bool }, defaults: { <opt>: str } }
// (see wyrd/feature_flags.py). This module owns the SPA half: the canonical
// flag NAMES, the flag→field mapping, and the english-always rule. Resolution
// is fail-closed — an option whose controlling flag isn't on is hidden, so
// default (no env set) = a minimal, prod-safe surface.
//
// The flags map is keyed by the env-var suffix the server saw (e.g.
// WYRD_FF_CULTURE_WELSH → "CULTURE_WELSH"); we compute the same suffix from a
// flag name via envify(), so namespaced flags round-trip without the server
// having to reverse-parse 'CULTURE_WELSH' back to 'culture.welsh'.

// Cultures always available regardless of flags — the generator needs a
// culture, and english is the guaranteed baseline.
const ALWAYS_ON_CULTURES = new Set(['english']);

// Advanced fields that are unconditionally available regardless of any flag.
// The vector axis-weight knobs are always on now that vector is the only
// scoring path (the scoring_mode selector was retired): there's no mode to
// gate them under, so they ride alongside their always-present sliders.
const ALWAYS_ON_FIELDS = new Set([
  'phonological_weight',
  'semantic_weight',
  'position_weight',
  'baseline_weight',
]);

// Advanced fields that ride under another field's flag rather than their own.
// (None currently — kept as the extension point for grouped gating.)
const FIELD_FLAG_OVERRIDES = {};

/** The flag name controlling an advanced field. Defaults to the field key
 *  itself (1:1); a few fields map to a grouping flag. */
export function fieldFlag(fieldKey) {
  return FIELD_FLAG_OVERRIDES[fieldKey] || fieldKey;
}

/** Envify a flag name into the manifest flags-map key (matches the server's
 *  WYRD_FF_<NAME> suffix): upper-case, '.' and '-' → '_'. */
export function envify(name) {
  return name.toUpperCase().replace(/[.-]/g, '_');
}

/** Resolve a flag against the manifest config. The master override (`all`)
 *  wins; an absent flag is off (fail-closed). `config` may be null/undefined
 *  (a legacy manifest with no config block) → everything off. */
export function flagOn(config, name) {
  if (!config) return false;
  if (config.all) return true;
  return config.flags?.[envify(name)] === true;
}

/** Is an advanced field enabled? Some fields are unconditionally available
 *  (always-on); the rest gate on their controlling flag. */
export function fieldEnabled(config, fieldKey) {
  if (ALWAYS_ON_FIELDS.has(fieldKey)) return true;
  if (!config) return false;
  return flagOn(config, fieldFlag(fieldKey));
}

/** Filter a culture enum down to english (always) + flagged-on cultures,
 *  preserving the schema's order. */
export function visibleCultures(config, cultureEnum) {
  return (cultureEnum || []).filter(
    (c) => ALWAYS_ON_CULTURES.has(c) || flagOn(config, `culture.${c}`),
  );
}

/** Coerce an env-string default override to the field's schema type (env
 *  values arrive as strings; the schema default is already typed). Array
 *  fields can't be seeded from a scalar env string, so they return undefined
 *  (caller falls back to the schema default / empty array) rather than
 *  seeding a raw string into an array slot. */
export function coerceToType(raw, prop) {
  if (prop?.type === 'array') return undefined;
  if (prop?.type === 'integer' || prop?.type === 'number') {
    // A bad/empty numeric env (e.g. WYRD_DEFAULT_COUNT=abc, or '') would
    // otherwise yield NaN / 0 and silently seed a junk value. Reject anything
    // NON-FINITE, not just NaN: Number('Infinity') and an overflow like
    // Number('1e999') both yield Infinity, which Number.isNaN does NOT catch
    // and which serializes to null in the request — the very junk this guard
    // exists to stop. Number.isFinite rejects NaN AND ±Infinity, matching the
    // server's math.isfinite mirror (feature_flags._coerce_to_schema_type).
    // Return undefined so the caller falls back to the schema default rather
    // than shipping garbage.
    if (String(raw).trim() === '') return undefined;
    const n = Number(raw);
    if (prop.type === 'integer') {
      // Integer parity with the server (_coerce_to_schema_type, wyrd-8uuq):
      // accept integral floats ('5.0'→5), reject non-integers and trailing
      // garbage ('5.5'/'3abc'/'12px'→undefined). Number.isInteger is false for
      // NaN and ±Infinity, so it subsumes the non-finite guard on this path.
      // (Was parseInt(raw,10), which leniently parsed '3abc'→3 — a slider≠
      // generation divergence against the server's strict int(); wyrd-8uuq.)
      return Number.isInteger(n) ? n : undefined;
    }
    return Number.isFinite(n) ? n : undefined;
  }
  if (prop?.type === 'boolean') {
    // .trim() for parity with the server (feature_flags._parse_bool strips) and
    // with this function's own number branch (String(raw).trim()): a padded
    // ' true ' must coerce to true, not silently false.
    return ['1', 'true', 'yes', 'on'].includes(String(raw).trim().toLowerCase());
  }
  return raw;
}

/** The seed value for a field: env default-override (coerced) wins over the
 *  schema default. Returns `undefined` when neither is set (caller applies
 *  its type-based empty fallback). */
export function seedDefault(config, fieldKey, prop) {
  const override = config?.defaults?.[fieldKey];
  if (override !== undefined) {
    const coerced = coerceToType(override, prop);
    // A junk override (coerced → undefined, e.g. WYRD_DEFAULT_COUNT=abc)
    // falls through to the schema default rather than seeding nothing.
    if (coerced !== undefined) return coerced;
  }
  return prop?.default;
}

/** wyrd-etvd: snap-to-valid for an enum/option select. Returns the value the
 *  field SHOULD be set to, or `undefined` to mean "leave it alone".
 *
 *  Critically, an `undefined` (not-yet-seeded) value returns `undefined` — the
 *  seed pass (``seedDefault``) owns initial seeding, including any env
 *  ``config.defaults`` override (e.g. WYRD_DEFAULT_COUNT=5).
 *  Snapping an unseeded field to ``schemaDefault`` here would clobber that
 *  override and the seed pass would then skip (value no longer undefined).
 *  Only a DEFINED-but-invalid value (e.g. a culture-filtered option) snaps —
 *  to the schema default if it's valid, else the first option.
 *
 *  @param value the current field value (may be undefined)
 *  @param options the valid options (prop.enum, or culture-filtered options)
 *  @param schemaDefault the schema's default value (prop.default) */
export function snapEnumValue(value, options, schemaDefault) {
  if (value === undefined) return undefined;
  if (options.includes(value)) return undefined;
  return schemaDefault !== undefined && options.includes(schemaDefault)
    ? schemaDefault
    : options[0];
}

/** wyrd-kqyf: the culture-agnostic era-default token (WYRD_DEFAULT_ERA /
 *  terraform feature_flag_defaults). Era stages are per-culture, so the env
 *  default can't name a literal stage; the token means "this culture's
 *  present-day stage" on both sides of the contract (the server translates it
 *  in _resolve_era_param / _resolve_era_render_language). */
export const PRESENT_DAY_ERA = 'present-day';

/** wyrd-kqyf: snap-to-valid for a DEPENDENT select (x-options-by-culture),
 *  layering the present-day translation on snapEnumValue:
 *
 *  - a seeded `present-day` value resolves to the culture's present-day stage
 *    — the LAST option (per-culture lists are chronologically ordered with ''
 *    first), so the env default displays as e.g. "Modern English" instead of
 *    binding an invalid select value;
 *  - when the CONFIG default is the token, a culture switch that invalidates
 *    the prior stage (english's 'modern-english' isn't a welsh option) snaps
 *    to the NEW culture's present-day stage rather than clobbering to the
 *    schema default — without this the env default would silently become
 *    Mixed Era (and be sent as an explicit '') after every culture change.
 *
 *  Returns the value to set, or undefined for "leave it alone" (same contract
 *  as snapEnumValue). */
export function snapDependentValue(value, options, prop, config, fieldKey) {
  if (value === undefined) return undefined;
  if (!options || options.length === 0) return undefined;
  const presentDay = options[options.length - 1];
  if (value === PRESENT_DAY_ERA) return presentDay;
  const target =
    seedDefault(config, fieldKey, prop) === PRESENT_DAY_ERA ? presentDay : prop?.default;
  return snapEnumValue(value, options, target);
}

/** wyrd-b6hd: the value a field should be INITIALIZED to — the env
 *  default-override (config.defaults) or schema default, falling back to a
 *  type-appropriate empty when neither is set. The store seeds every field
 *  with this up front (appState.ensureParams), so the form binds populated
 *  values and no per-component lazy seed races the <select> bind (wyrd-etvd).
 *  Always returns a defined value (never undefined) so the bound state is
 *  never empty at mount. */
export function initialFieldValue(config, fieldKey, prop) {
  const seed = seedDefault(config, fieldKey, prop);
  if (seed !== undefined) return seed;
  if (prop?.type === 'array') return [];
  if (prop?.type === 'boolean') return false;
  return '';
}

File: wyrd/feature_flags.py
"""Feature-flag + default-value resolution for the SPA config UI (wyrd-0gou).

The SPA gates each advanced configuration option behind a feature flag so
untested options stay hidden in production and can be turned on one-by-one as
they're validated. This is a UI concern only — the API stays permissive; the
flags only decide what the SPA *renders*.

Resolution is env-driven, so flags + default values are settable per
environment with no rebuild (the resolved config rides on /api/manifest):

  WYRD_FF_ALL=true            master override → every flag on (staging sets
                              this so a forgotten flag can't hide an option)
  WYRD_FF_<NAME>=true|false   per-flag override; NAME is the flag name
                              upper-cased with '.' / '-' replaced by '_'
                              (e.g. flag ``culture.welsh`` → WYRD_FF_CULTURE_WELSH)
  WYRD_DEFAULT_<OPTION>=<v>   override an option's default VALUE (e.g.
                              WYRD_DEFAULT_CULTURE=english)

Every flag defaults OFF. The shipped shape is

  {"all": bool, "flags": {<ENVIFIED_NAME>: bool}, "defaults": {<option>: str}}

and the SPA resolves a flag as ``all OR flags[envify(name)]`` (absent → off).
The canonical flag-name list lives on the SPA side (it owns the flag→field
mapping); the server stays registry-free, so adding a flag is just a new env
var + a new SPA mapping entry — no change here.
"""

from __future__ import annotations

import math
import os
import re
from collections.abc import Mapping

# Leading optional-sign integer, mirroring JS ``parseInt(raw, 10)``: it parses
# the leading run of digits and ignores any trailing fractional/garbage part
# (``parseInt('5.0', 10) === 5``, ``parseInt('5px', 10) === 5``). ASCII ``[0-9]``,
# NOT ``\d``: ``\d`` also matches Unicode digits (Arabic-Indic ``١٢٣``, Devanagari)
# which ``int()`` would then accept, but JS ``parseInt`` is ASCII-only and yields
# NaN there — so ``\d`` would re-introduce a server/SPA divergence, not close one.
_LEADING_INT_RE = re.compile(r"[+-]?[0-9]+")

_FF_PREFIX = "WYRD_FF_"
_DEFAULT_PREFIX = "WYRD_DEFAULT_"
_ALL_VAR = "WYRD_FF_ALL"
_TRUE_VALUES = frozenset({"1", "true", "yes", "on"})


def _parse_bool(raw: str) -> bool:
    return raw.strip().lower() in _TRUE_VALUES


def resolve_feature_config(env: Mapping[str, str] | None = None) -> dict:
    """Resolve the SPA feature-flag + default-value config from ``env``.

    Returns ``{"all": bool, "flags": {envified_name: bool}, "defaults":
    {option: str}}``:

    * ``all`` — the master override (``WYRD_FF_ALL``). When true the SPA
      treats every flag as on regardless of individual settings.
    * ``flags`` — every explicitly-set ``WYRD_FF_<NAME>`` (excluding
      ``WYRD_FF_ALL``), keyed by its envified suffix, parsed to bool.
    * ``defaults`` — every ``WYRD_DEFAULT_<OPTION>``, keyed by the
      lower-cased option name, value kept as the raw string (the SPA
      coerces to the field's schema type when seeding).

    Pure: reads only the passed mapping (``os.environ`` by default), so it
    is trivially testable with a synthetic env.
    """
    src = os.environ if env is None else env
    all_on = _parse_bool(src.get(_ALL_VAR, ""))
    flags: dict[str, bool] = {}
    defaults: dict[str, str] = {}
    for key, value in src.items():
        if key == _ALL_VAR:
            continue
        if key.startswith(_FF_PREFIX):
            # Keep the envified suffix as the key; the SPA computes the same
            # suffix from the flag name, so no ambiguous reverse-parsing of
            # 'CULTURE_WELSH' → 'culture.welsh' is ever needed.
            flags[key[len(_FF_PREFIX) :]] = _parse_bool(value)
        elif key.startswith(_DEFAULT_PREFIX):
            defaults[key[len(_DEFAULT_PREFIX) :].lower()] = value
    return {"all": all_on, "flags": flags, "defaults": defaults}


def _coerce_to_schema_type(raw: str, prop: Mapping) -> object | None:
    """Coerce a ``WYRD_DEFAULT_*`` string to an option's JSON-schema type.

    The server-side mirror of the SPA's ``featureFlags.coerceToType``. Returns
    ``None`` for an empty/uncoercible value or an ``array`` option, so the
    caller falls back to the option's own schema default rather than seeding
    garbage (matches the SPA, which returns ``undefined`` in those cases)."""
    schema_type = (prop or {}).get("type")
    if schema_type == "array":
        return None
    if schema_type == "integer":
        # Mirror the SPA's ``parseInt(raw, 10)``: take the leading optional-sign
        # integer, ignoring a trailing fractional/garbage part ("5.0"→5, "5px"→5).
        # Plain ``int(text)`` rejects those (ValueError) and falls back to the
        # schema default while the SPA seeds the parsed number — exactly the
        # wyrd-7f22 SPA/server mismatch this module exists to prevent.
        m = _LEADING_INT_RE.match(str(raw).strip())
        return int(m.group()) if m is not None else None
    if schema_type == "number":
        text = str(raw).strip()
        if not text:
            return None
        try:
            value = float(text)
        except ValueError:
            return None
        # float() accepts "nan"/"inf"; reject non-finite so an operator typo
        # (WYRD_DEFAULT_NOVELTY=inf) can't poison the score blend or serialize as
        # invalid-JSON `NaN`/`Infinity` in the echoed params. Mirrors the SPA's
        # Number.isFinite guard (featureFlags.coerceToType — which also rejects an
        # overflow like Number('1e999') → Infinity).
        if not math.isfinite(value):
            return None
        return value
    if schema_type == "boolean":
        return _parse_bool(str(raw))
    return raw


def apply_env_defaults(
    params: dict, properties: Mapping, env: Mapping[str, str] | None = None
) -> dict:
    """Fill option values from ``WYRD_DEFAULT_<OPTION>`` for options ABSENT from
    a generation request, coerced to each option's schema type. Mutates and
    returns ``params``.

    This is the server-side half of the ``WYRD_DEFAULT`` contract (CLAUDE.md:
    it overrides an option's default *value*, not just its SPA visibility). The
    SPA seeds the same overrides for display, but it omits any field still at
    its seeded default from the request (the wyrd-14hn "default = don't include"
    rule), so without this the server would fall back to the SCHEMA default and
    a value-changing override like ``WYRD_DEFAULT_NOVELTY=0.1`` would be a no-op
    on generation while the slider showed 0.1 (wyrd-7f22). Only fills keys that
    are real options for this generator and absent from the request, so an
    explicit request value (incl. one equal to the schema default) always wins
    and bit-stability holds when no ``WYRD_DEFAULT_*`` is set."""
    defaults = resolve_feature_config(env)["defaults"]
    for key, raw in defaults.items():
        if key in params:
            continue
        prop = properties.get(key)
        if prop is None:
            continue
        coerced = _coerce_to_schema_type(raw, prop)
        if coerced is not None:
            params[key] = coerced
    return params
